leaderelection/context.go

```diff
@@ -0,0 +1,218 @@
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+	"knative.dev/pkg/system"
+)
+
+// WithStandardLeaderElectorBuilder infuses a context with the ability to build
+// LeaderElectors with the provided component configuration acquiring resource
+// locks via the provided kubernetes client.
+func WithStandardLeaderElectorBuilder(ctx context.Context, kc kubernetes.Interface, cc ComponentConfig) context.Context {
+	return context.WithValue(ctx, builderKey{}, &standardBuilder{
+		kc:  kc,
+		lec: cc,
+	})
+}
+
+// HasLeaderElection returns whether there is leader election configuration
+// associated with the context
+func HasLeaderElection(ctx context.Context) bool {
+	val := ctx.Value(builderKey{})
+	return val != nil
+}
+
+// Elector is the interface for running a leader elector.
+type Elector interface {
+	Run(context.Context)
+}
+
+// BuildElector builds a leaderelection.LeaderElector for the named LeaderAware
+// reconciler using a builder added to the context via WithStandardLeaderElectorBuilder.
+func BuildElector(ctx context.Context, la reconciler.LeaderAware, name string, enq func(reconciler.Bucket, types.NamespacedName)) (Elector, error) {
+	if val := ctx.Value(builderKey{}); val != nil {
+		switch builder := val.(type) {
+		case *standardBuilder:
+			return builder.BuildElector(ctx, la, name, enq)
+		}
+		// TODO(mattmoor): Add a flavor of builder that relies on StatefulSet to partition the key space.
+	}
+
+	return &unopposedElector{
+		la:  la,
+		bkt: reconciler.UniversalBucket(),
+		enq: enq,
+	}, nil
+}
+
+type builderKey struct{}
+
+type standardBuilder struct {
+	kc  kubernetes.Interface
+	lec ComponentConfig
+}
+
+func (b *standardBuilder) BuildElector(ctx context.Context, la reconciler.LeaderAware, name string, enq func(reconciler.Bucket, types.NamespacedName)) (Elector, error) {
+	logger := logging.FromContext(ctx)
+
+	id, err := UniqueID()
+	if err != nil {
+		return nil, err
+	}
+
+	buckets := make([]Elector, 0, b.lec.Buckets)
+	for i := uint32(0); i < b.lec.Buckets; i++ {
+		bkt := &bucket{
+			component: b.lec.Component,
+			name:      name,
+			index:     i,
+			total:     b.lec.Buckets,
+		}
+
+		rl, err := resourcelock.New(b.lec.ResourceLock,
+			system.Namespace(), // use namespace we are running in
+			bkt.Name(),
+			b.kc.CoreV1(),
+			b.kc.CoordinationV1(),
+			resourcelock.ResourceLockConfig{
+				Identity: id,
+			})
+		if err != nil {
+			return nil, err
+		}
+		logger.Infof("%s will run in leader-elected mode with id %q", bkt.Name(), rl.Identity())
+
+		le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: b.lec.LeaseDuration,
+			RenewDeadline: b.lec.RenewDeadline,
+			RetryPeriod:   b.lec.RetryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(context.Context) {
+					logger.Infof("%q has started leading %q", rl.Identity(), bkt.Name())
+					la.Promote(bkt, enq)
+				},
+				OnStoppedLeading: func() {
```

Comment:
1. The only return path from `Run` is [here](https://github.com/kubernetes/client-go/blob/18c1ef0354fe489b5d9a58088676f854be243aa2/tools/leaderelection/leaderelection.go#L204) which I understood to only happen when context is cancelled.  This should also prevent iteration in the loop below.

2. My understanding is that `defer runtime.HandleCrash()` is intended to keep panics from killing things.

3. Yes, but the `le.renew` call is not and blocks the return.

Reply:
>My understanding is that defer runtime.HandleCrash() is intended to keep panics from killing things.

It runs handlers and then rethrows  - I'm surprised it's in that Run method

https://github.com/kubernetes/apimachinery/blob/v0.17.6/pkg/util/runtime/runtime.go#L55

Reply:
circling back for others

1)  is inaccurate & @mattmoor is correct - so the tight loop is in the leader election code in client-go

2) because 1 was inaccurate this means 2 (`OnStoppedLeading blocks LeaderElector.Run`) isn't a 'big' problem

3) my intent here was to highlight the difference on how the callback `Promote` & `Demote` behave different (since one blocks Run and the other doesn't). It's not a huge issue because of my bad assumption in 1)